fast64_internal/oot/exporter/cutscene/__init__.py

```diff
@@ -0,0 +1,145 @@
+import bpy
+
+from dataclasses import dataclass, field
+from typing import Optional
+from bpy.types import Object
+from ....utility import PluginError, CData, indent
+from ...oot_utility import getCustomProperty
+from ...scene.properties import OOTSceneHeaderProperty
+from ..base import Base
+from .data import CutsceneData
+
+
+# NOTE: ``paramNumber`` is the expected number of parameters inside the parsed commands,
+# this account for the unused parameters. Every classes are based on the commands arguments from ``z64cutscene_commands.h``
+
+# NOTE: ``params`` is the list of parsed parameters, it can't be ``None`` if we're importing a scene,
+# when it's ``None`` it will get the data from the cutscene objects
+
+
+@dataclass
+class Cutscene:
+    """This class defines a cutscene, including its data and its informations"""
+
+    csObj: Object
```

Comment:
I don't like using ``__init__()``, I think when you have args it makes the code not pleasant to read, that's part of the reason why I used dataclasses

I know it's too bad to use ``__post_init__`` but idk I just don't think code like this:
```py
def __init__(arg1, arg2):
    self.arg1 = arg1
    self.arg2 = arg2
```
is aesthetic at all, also using dataclasses makes it obvious which type of class you deal with imo (you could say every class holds data but in this context I mean the data related to the exported scene)

I guess I just misinterpreted the official doc then but I thought it was something there just in case you needed to init more things, not a fan of ``__init__()`` but if it blocks the merge I'll make the change

Reply:
Dataclasses are definitely useful when your `__init__()` takes in all the data fields as arguments, but a lot of the classes here instead take things like blender objects/properties and then fill out the fields from those objects. Having all the data field arguments be optional just to make the dataclass work doesn't make sense, because someone could technically pass in those optional arguments, but have the result not be what they expect since those arguments get overridden.

The dataclass documentation says you can use `InitVar` for init only variables, but that doesn't really apply here, since none of the optional fields are really optional - you specifically only want them to get their values from other objects you pass in. `__post_init__()` can be used to initialize other derived fields, but you would want to use `field(init=False)` in that case, and even then the documentation itself states that its preferable to just define your own `__init__()` function. This feels a bit like a square peg in a round hole situation.

I don't think classes should be annotated with `@dataclass` solely for documentation purposes, but if you want to keep it, you should define a custom `__init__()` at least. Or, you could create a `@staticmethod` that creates the dataclass object based on only the necessary arguments, although this is basically the same thing as an `__init__()`.

Reply:
I'm not sure how to solve this but it sounds to me like the builder pattern (more or less) may be an idea here?

Like you create an object with nothing and iteratively .setThing / .withThing on it (maybe it checks some things don't overwrite previous data too).
Then at the end you .build and it checks everything is there and builds a dataclass with all the things collected

However the builder pattern is not pythonic at all, I've never seen it in python code

Reply:
I looked at it again and I don't see how any of the parameters are overriden by post init, none of the arguments you pass when calling the class are immediately changed, in fact they never get changed, you're using the parameters to set other class members that don't have a default value, which is the purpose of post init according to the [docs](https://docs.python.org/3.10/library/dataclasses.html#post-init-processing):
```py
@dataclass
class C:
    a: float
    b: float
    c: float = field(init=False)

    def __post_init__(self):
        self.c = self.a + self.b
```
this is exactly what I do basically everywhere here, my only mistake to my knowledge is doing ``c: Optional[float] = None`` instead of ``c: float = field(init=False)``, so this is the only change I'll agree on unless there's something I completely misunderstood about dataclasses

Reply:
```py
@dataclass
class Cutscene:
    """This class defines a cutscene, including its data and its informations"""

    csObj: Object
    useMacros: bool
    name: Optional[str] = None
    motionOnly: bool = False
    data: Optional[CutsceneData] = None
    totalEntries: int = 0
    frameCount: int = 0
    paramNumber: int = 2

    def __post_init__(self):
        # when csObj is None it means we're in import context
        if self.csObj is not None and self.data is None:
            if self.name is None:
                self.name = self.csObj.name.removeprefix("Cutscene.").replace(".", "_")
            self.data = CutsceneData(self.csObj, self.useMacros, self.motionOnly)
            self.totalEntries = self.data.totalEntries
            self.frameCount = self.data.frameCount
```
calling this class looks like this: ``Cutscene(csObj, True)``, I don't see how these are overriden in the post init, unless you meant the other members but that would fall under the mistake I mentioned earlier (which would solve your other concern: https://github.com/Fast-64/fast64/pull/256#discussion_r1468267514 I think)